### ISSUE TITLE
Incorrect validator number distribution function

### ISSUE THREAD

**Issue by r1d1-btct** [semuxproject/semux#620]
Hi!

#### Expected behaviour

The Semux client should choose validator to forge a block based on uniform distribution (so any validator on a long period of time should forge so many blocks as other).

#### Actual behaviour

Validators with numbes 1-56 has an advantage over others. The difference is about 1.5x
For example, if validator with number 1-56 will forge 150 blocks, others will forge only 100 blocks each.

#### Steps to reproduce the behaviour

I found this bug when my validator, which is constantly at the bottom of the table, forged significantly fewer blocks than it should be by calculation.

The problem is: 

	validators.get((Hash.h256(key)[0] & 0xff) % validators.size());

h256 is uniform on 0->255, but after %100 we get not uniform distibution. 
Cause first 200 give a uniform distribution, but the additional 55 corrupt everything.

For a visual demonstration, I plotted a graph showing the problem. 
The series2 is result of working h256 function, the series1 is result of h256 % 100.
![distribution](https://user-images.githubusercontent.com/32393567/35607766-6e390ca6-0688-11e8-9b1d-9a9ee1d4e580.png)

#### Solution

I think you should use something like this: 

	long seed = new BigInteger(key).longValue();
	int number = new Random(seed).nextDouble() * 100;

This will give a uniform distribution and it's gives the same values for the same sid (height and view).

But I don't know this java Random implementation is finish or not. I.e., whether it can changed in the future or not.
Perhaps it's safer to use its implementation:

	public static int random100(long seed) {
		long oldSeed, nextSeed;
		AtomicLong s = new AtomicLong(seed);
		do {
			oldSeed = s.get();
			nextSeed = (oldSeed * 0x5DEECE66DL + 0xBL) & ((1L << 48) - 1);
		} while (!s.compareAndSet(oldSeed, nextSeed));
		int nextInt = (int)(nextSeed >>> (48 - 24));
		return (int) (nextInt / (float)(1 << 24) * 100);
	}

Returns [0-100).
It's based on default Java implemantion. 
I tested this it's give a uniform distrubion as expected.
The deviation of distribution is about 0.2%.
But I did not test this solution very long. Probably I did not take something into account. 
And maybe there is another simpler method of solving the problem.
And as I understand we need hardwork for it and you should save old function before hardfork block.

#### Test
		
	public static void main(String[] args) {
		byte []key;
		Random random = new Random();
		
		HashMap<Integer, AtomicInteger> map100 = new HashMap<>();
		HashMap<Integer, AtomicInteger> mapPRNG = new HashMap<>();
		HashMap<Integer, AtomicInteger> map256 = new HashMap<>();
		
		for (int i = 0; i < 256; i++) {
			map256.put(i, new AtomicInteger(0));
			if(i < 100) {
				map100.put(i, new AtomicInteger(0));
				mapPRNG.put(i, new AtomicInteger(0));
			}
		}

		int view, v256, v100, vPNRG;
		long seed;
		for (int i = 0; i < 1_000_000; i++) {
			view = random.nextDouble() < 0.05 ? 1: 0; //about 5%
			key = Bytes.merge(Bytes.of(i), Bytes.of(view));
			v256 = Hash.h256(key)[0] & 0xff;
			v100 = v256 % 100;

			seed = new BigInteger(Bytes.merge(Bytes.of(i), Bytes.of(view))).longValue();
			//vPNRG = (int)(new Random(seed).nextDouble() * 100);
			vPNRG = random100(seed);

			map100.get(v100).incrementAndGet();
			map256.get(v256).incrementAndGet();
			mapPRNG.get(vPNRG).incrementAndGet();
		}
		System.out.println("Number\th256\t%100\tPRNG");
		for (int i = 0; i < 256; i++) {
			System.out.println(i + "\t" + map256.get(i).intValue() +
					"\t" + (map100.containsKey(i)? map100.get(i).intValue(): 0) +
					"\t" + (mapPRNG.containsKey(i)? mapPRNG.get(i).intValue(): 0)
			);
		}
	}

	public static int random100(long seed) {
		long oldSeed, nextSeed;
		AtomicLong s = new AtomicLong(seed);
		do {
			oldSeed = s.get();
			nextSeed = (oldSeed * 0x5DEECE66DL + 0xBL) & ((1L << 48) - 1);
		} while (!s.compareAndSet(oldSeed, nextSeed));
		int nextInt = (int)(nextSeed >>> (48 - 24));
		return (int) (nextInt / (float)(1 << 24) * 100);
	}

**Comment by orogvany** [comment]
I saw this and thought it was some intentional thing that higher voted validators Got more blocks.  ;)

**Comment by semuxdev** [comment]
I don't think we can use random generator for validator selection, assuming it's deterministic.  One validator can keep the `primary` status forever by making proper block headers and precomputing the next validator.

**Comment by semuxdev** [comment]
It was not intended behaviour, but may be nice to see tiered delegate payout.

Also the random generation process can be simplified into
```
new BigInteger(1, key).mod(BigInteger.valueOf(validators.size()).intValue()
```

**Comment by cryptokat** [comment]
Hi @semuxdev 
I've surveyed the source code in JDK regarding mentioned BigInteger initializer. It appears to be opening a possibility to `ArtihmeticException` throwing from `BigInteger#reportOverflow`.

References:
- https://github.com/openjdk-mirror/jdk/blob/adea42765ae4e7117c3f0e2d618d5e6aed44ced2/src/share/classes/java/math/BigInteger.java#L345-L361
- https://github.com/openjdk-mirror/jdk/blob/adea42765ae4e7117c3f0e2d618d5e6aed44ced2/src/share/classes/java/math/BigInteger.java#L1068-L1072

**Comment by semuxdev** [comment]
BigInteger should be able to handle 256-bit arithmetic operations easily. The max magnitude is '1 << 26', which is far beyond is the 32 bytes.

**Comment by cryptokat** [comment]
@semuxdev 
I created a validation script based on @r1d1-btct 's work. The initializer of `BigInteger(int signum, byte[] magnitude)` doesn't seem to be well-distributed either.

https://gist.github.com/cryptokat/e5d130364c5db50b255201c93481fef5
```
Number	h256	%100	PRNG	BigInteger
0	3906	11795	10017	37975
1	3921	11586	9995	2025
2	3920	11711	10005	0
3	3880	11690	9997	0
4	3827	11581	9995	37933
5	3946	11797	10008	2067
6	3807	11495	9991	0
7	3840	11651	9995	0
8	3914	11645	10007	37927
9	3827	11566	9998	2073
10	3956	11754	10000	0
11	3802	11570	10002	0
12	3879	11754	10004	37966
13	3799	11688	10005	2034
14	3893	11713	9994	0
15	3803	11701	9991	0
16	3852	11837	10010	38003
17	3908	11691	9996	1997
18	3939	11641	9993	0
19	3860	11712	10012	0
20	3901	11717	9999	37960
21	4015	11873	9997	2040
22	3872	11612	10003	0
23	3955	11856	9996	0
24	3924	11734	9994	37988
25	3915	11678	10011	2012
26	3939	11813	9989	0
27	3943	11710	10014	0
28	3852	11650	9993	38060
29	3909	11631	9995	1940
30	4028	11904	10019	0
31	3903	11793	9990	0
32	3969	11806	9996	38012
33	3953	11802	10005	1988
34	3723	11553	9993	0
35	3905	11782	9994	0
36	3844	11711	10009	38011
37	3936	11819	9995	1989
38	3907	11675	10010	0
39	3812	11759	9988	0
40	3891	11648	10008	37950
41	3958	11869	10009	2050
42	3822	11656	9988	0
43	3908	11824	9997	0
44	3980	11994	10007	38045
45	3935	11736	9994	1955
46	3890	11589	9995	0
47	3956	11714	10007	0
48	3973	11842	9993	37978
49	3886	11718	9997	2022
50	3802	11613	10015	0
51	3941	11630	9991	0
52	3934	11779	10017	37972
53	3934	11749	9990	2028
54	3879	11645	9996	0
55	3923	11687	10002	0
56	3903	7735	10001	38016
57	3800	7779	9989	1984
58	3906	7885	10010	0
59	3919	7821	9989	0
60	3940	7875	10009	37998
61	3951	7845	10005	2002
62	3802	7778	9995	0
63	3973	7942	10006	0
64	3856	7744	10000	38000
65	4049	8022	9994	2000
66	3998	7863	10008	0
67	3893	7662	9993	0
68	3974	7851	9993	37917
69	3930	7753	10009	2083
70	3978	7766	9997	0
71	3874	7788	9998	0
72	3943	7771	10009	37992
73	3964	7875	9993	2008
74	3996	7921	9995	0
75	3776	7548	10010	0
76	3887	7845	9993	38024
77	3892	7776	10010	1976
78	3770	7705	9992	0
79	3950	7829	9989	0
80	3860	7720	10014	37940
81	3942	7905	10000	2060
82	3894	7887	9996	0
83	3991	7851	10009	0
84	3902	7823	9991	38003
85	3905	7908	9997	1997
86	3965	7891	10008	0
87	3882	7783	9992	0
88	3871	7826	10008	37966
89	3909	7786	9995	2034
90	3982	7881	10000	0
91	3883	7764	10013	0
92	3822	7811	9992	38014
93	4002	7985	9996	1986
94	3970	7928	10005	0
95	3820	7645	9999	0
96	3862	7805	9993	37991
97	3766	7663	10009	2009
98	3834	7724	9995	0
99	3975	7886	9990	0
```

**Comment by semuxdev** [comment]
Sorry, I should've given you more definition. The key I wrote before is supposed to be the hash of height and view. I'm assuming Blake2b hash will generate better distribution than the random generator.

**Comment by r1d1-btct** [comment]
I did some additional tests. The random generator is a very bad choice on a small distance.

**Comment by orogvany** [comment]
You could possibly keep blake and just fix the 255 issue. If it is evenly distributed either scale it 1-100 or discard and regen above that value.  Will need modification if/when we get past 255 validators though

**Comment by orogvany** [comment]
Or maybe thats what youre trying already

**Comment by cryptokat** [comment]
Added a round-robin function:
```
BigInteger.valueOf(i + random100(view)).mod(BigInteger.valueOf(100)).intValue()
```

https://gist.github.com/cryptokat/8225280d1fcff34a4b7eadeeeadc872e
![selection_217](https://user-images.githubusercontent.com/32390172/35857654-b4f70da0-0b74-11e8-8cbc-7980c0416057.png)

**Comment by semuxdev** [comment]
We were originally using round robin. The problem is 'h + 1 = (h + 1) + 0'. Validators generate consequent blocks. 

This is why we introduced hash to the validator selection.

BigInteger.valueOf(h256(Bytes.merge(Bytes.of(height), Bytes.of(view))).mod

**Comment by mdodong** [comment]
Why not _randomly shuffle_ the **validatorArrayList** at the start of each round? 

This may be useful: [StackOverflow-random-shuffling-of-an-array](https://stackoverflow.com/questions/1519736/random-shuffling-of-an-array)

**Comment by orogvany** [comment]
One would need to have a deterministic shuffle for this to work @mdodong as all validators would need to have the same shuffled order.  I'm not sure which implementations offer a shuffle seed, but it certainly is an option or is doable.
It gets problematic if we do change to more than 100 validators.  Right now it works out evenly to 2 blocks per round, but if we had 120 validators, and we did the shuffle, I don't know that we'd end up with equal distribution over time with this method.

**Comment by gpip** [comment]
There is a comment by @orogvany which is the common way to deal with this, you discard the result if it's outside the range. Everything stays the same except you drop the modulus, if it's above 100 you increment the view number and reapply the blake hash. Would it be ok to use the view param for this in addition to missed validations?

**Comment by r1d1-btct** [comment]
As I wrote [above ](https://github.com/semuxproject/semux/issues/620#issuecomment-363294810) the "random" is a bad choice in dinstance of 200 blocks. I must draw your attention to this one more time.

If the old version (round-robin) does not have problems, except consecutive blocks forging by one validator (if the previous validator missed its block), then I vote for it.
In extreme cases, you can mix the list of validators when saving (every 200 blocks) based on the hash of the block or on the hash of the current position of the delegates.

The problem with consecutive mining was inherent in the test network due to the small award and the low price of the coin, possibly due to some errors that were fixed. Some of the validators didn't work.
As far as I can see, there are much fewer missing validators now.

Check the work of the 'deterministicRandom' on a distance of 200 blocks and you will see that the good old RR is better.

**Comment by orogvany** [comment]
You are correct @r1d1-btct 
I'll need to refresh myself on why roundrobin was abandoned, it was a bit before my time, but we're aware of the shortcomings here, and trying to figure out what meets all needs.
thanks!

**Comment by mdodong** [comment]
@orogvany main issue with round robin is that the order of validation is fixed. The problem back then was with users running multiple validators piling them in consecutive order, then disabling some all but one to take the consecutive blocks in play. This allows them to forge multiple blocks and maintain only one healthy validator making the network weaker.

If we can avoid consecutive block forging and slash unhealthy or low performing validators then there is no problem with round robin.

**Comment by orogvany** [comment]
Ah, yes, we can certainly manage fixing that :D

**Comment by fransla** [comment]
Late to the party, but here goes:
PRNGs are tricky. The stream of random numbers has to be both random and uniformly distributed.
The proposed solution has neither of those properties. It is based on the java.util.Random (https://docs.oracle.com/javase/8/docs/api/java/util/Random.html) implementation, and as such, inherits its weak LCG PRNG, which was already leaking seed bits and could easily be "cracked" to retrieve the seed value with just a few outputs.
See the following references for attacks:
https://crypto.stackexchange.com/questions/2086/predicting-values-from-a-linear-congruential-generator
http://jazzy.id.au/default/2010/09/20/cracking_random_number_generators_part_1.html
The same faulty java.util.Random is also used in the array shuffle function here:
`org.semux.util.ArrayUtil.java`

Even worse, the LCG fails spectacularly when n is a power of 2, and although the java.util.Random adjusts for it, the suggested deterministicRandom() function does not - direct quote from JavaDocs:

> _The algorithm treats the case where n is a power of two specially: it returns the correct number of high-order bits from the underlying pseudo-random number generator. In the absence of special treatment, the correct number of low-order bits would be returned. Linear congruential pseudo-random number generators such as the one implemented by this class are known to have short periods in the sequence of values of their low-order bits. Thus, this special case greatly increases the length of the sequence of values returned by successive calls to this method if n is a small power of two._

For n=100 this will not be a problem, but if the number of validators are increased, a potential problem lurks at n=128.

Finally, the java.util.Random has a check to prevent uneven distribution by discarding the tail end of values (this clause):
` } while (bits - val + (bound-1) < 0)`
which is not present in deterministicRandom(). Granted, it is a small problem given the size of n versus the number of bits used (24 in this case), but still notable as the purpose of the fix is to get rid of bias.

A much simpler solution would be to apply a similar fix that the java.util.Random uses to prevent uneven distribution in the first place, to the original implementation:

```
int index = 0;
int chosen = -1;
int hashbyte = 0;

do {
	hashbyte = Hash.h256(key)[index] & 0xff;
	chosen = hashbyte % validators.size();
	index++;
} while (hashbyte - chosen + validators.size() - 1 > 255);
```
Here is the code @cryptokat used to test above, modified for the fix:
http://tpcg.io/JnhG3d

Sample output:
```
Number	h256	%100	PRNG
0	5006	9979	10016
1	4972	9959	9995
2	4898	9997	10007
3	4978	10077	9993
4	5025	10098	9995
5	5207	10205	10009
6	5056	10089	9996
7	5057	10082	9994
8	5026	9984	10005
9	4944	9924	9996
10	5014	10018	10005
11	4928	9946	10004
12	5101	10178	9998
13	5000	10131	10004
14	5080	10121	9998
15	4973	9922	9990
16	4900	9931	10010
17	4938	9990	9991
18	5068	9985	9999
19	5047	10090	10007
20	5047	10036	10001
21	4979	9807	9995
22	5005	10045	10002
23	5045	10051	10000
24	5000	10070	9998
25	4929	9942	9997
26	5021	9950	10000
27	5054	10033	10008
28	5025	9983	9997
29	5098	10103	9993
30	5138	10184	10017
31	4861	9876	9993
32	4917	10061	9993
33	5006	9986	10010
34	4986	9978	9994
35	4996	10029	9996
36	5079	9967	10000
37	4989	10050	10000
38	5004	9972	10008
39	4935	9866	9990
40	4973	10051	10004
41	4950	9954	10012
42	5041	9985	9990
43	4883	9881	9999
44	4950	9968	10008
45	5088	9982	9991
46	5052	10025	9994
47	4989	10018	10010
48	5069	9960	9989
49	4944	9868	9995
50	5033	9952	10020
51	5072	9989	9993
52	5050	10070	10006
53	4972	9971	9998
54	4984	9946	9988
55	5073	9956	10013
56	5046	9997	9996
57	5036	10097	9993
58	5011	10050	10011
59	4932	10029	9988
60	4951	9792	10005
61	5155	10203	10011
62	4933	9981	9991
63	5123	10143	10007
64	4925	9919	9996
65	5018	10071	9995
66	4964	10013	10006
67	4876	9778	9995
68	5028	10028	9991
69	5083	10028	10009
70	4908	9849	10003
71	5030	9923	9994
72	4942	9867	10009
73	5022	10138	9989
74	4992	9957	9998
75	4878	9854	10009
76	5034	9989	9994
77	4960	9978	10012
78	5014	10006	9995
79	4819	9790	9987
80	4947	9989	10022
81	5124	10077	9990
82	4981	10019	9997
83	5066	10129	10007
84	5049	10073	9998
85	5025	10165	9991
86	4995	9985	10010
87	4976	9836	9990
88	5100	10135	10012
89	4771	9727	9995
90	4874	9839	10000
91	5040	10090	10008
92	4982	10024	9996
93	4919	9811	9992
94	5094	9959	10007
95	4938	10075	9996
96	5015	10143	9993
97	5025	10057	10009
98	5024	10189	9993
99	4921	9927	9996
```

Of course, that would still not fix the 
`org.semux.util.ArrayUtil.java`
problem, but that can be logged as a separate issue.

Bug bounty welcome :-)

**Comment by cryptokat** [comment]
Hello @fransla 
Thank you for your valuable feedback :+1: 

Since there was an user trying to exploit the reward system by moving itself to the position which can make the highest rewards of each round (every 200 blocks), we've decided to use the simplest round-robin system + randomized backups such that users will no longer be able to make an advantage with the same trick.

Please check the latest implementation [here](https://github.com/semuxproject/semux/blob/fc70bfaf0c65fe3ea0f3f3aaa32001e2bda8e152/src/main/java/org/semux/config/AbstractConfig.java#L180) and [here](https://github.com/semuxproject/semux/blob/develop/src/main/java/org/semux/util/BigIntegerUtil.java#L127).

**Comment by cryptokat** [comment]
closed by #625

**Comment by fransla** [comment]
Cool, thanks.
